File: src/ZigZagConversion.java
/**
 * Created by Administrator on 2015/1/2.
 */
public class ZigZagConversion {
    public String convert(String s, int nRows){
        char[] arr = s.toCharArray();
        char[] res = new char[s.length()];
        int c = 0;
        int r = nRows * 2 - 2;
        if(r == 0){
            r = 1;
        }
        for(int i = 0; i < nRows; i += 1){
            for(int j = i; j < arr.length; j += r){
                res[c] = arr[j];
                c += 1;
                int next = j + 2 * (nRows - 1 - i);
                if(i != 0 && i != nRows - 1 && next < arr.length){
                    res[c] = arr[next];
                    c += 1;
                }
            }
        }
        return String.valueOf(res);
    }
    public static void test(){
//        System.out.println(new ZigZagConversion().convert("PAYPALISHIRING", 3));
        System.out.println(new ZigZagConversion().convert("P", 1));
    }
}

File: cpp/zigzag_conversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        vector<string> g;
        for(int i = 0; i < numRows; i += 1){
            g.push_back("");
        }
		if(numRows == 1){
			return s;
		}
        int d = 0;
        int inc = 1;
        int n = s.size();
        for(int i = 0; i < n; i += 1){
            g[d].push_back(s[i]);
            if(d == 0){
                inc = 1;
            }else if(d == numRows - 1){
                inc = -1;
            }
            d = d + inc;
        }
        string res = "";
        for(int i = 0; i < numRows; i += 1){
            res = res.append(g[i]);
        }
        return res;
    }
};